packages/monitor-v2/src/llm-bot/common.ts

```diff
@@ -0,0 +1,137 @@
+  requestTx: string; // Transaction hash of the request.
+  proposer?: string; // Address of the proposer.
+  proposedValue?: number | boolean; // Proposed value.
+  proposeTx?: string; // Transaction hash of the proposal.
+  disputableUntil?: number; // Timestamp in ms until the request can be disputed.
+  resolvedValue?: number | boolean; // Resolved value.
+  resolveTx?: string; // Transaction hash of the resolution.
+  disputeTx?: string; // Transaction hash of the dispute.
+
+  constructor(data: {
+    body: string;
+    type: OptimisticOracleType;
+    timestamp: number;
+    identifier: string;
+    requester: string;
+    requestTx: string;
+    proposer?: string;
+    proposedValue?: number | boolean;
+    proposeTx?: string;
+    disputableUntil?: number;
+    resolvedValue?: number | boolean;
+    resolveTx?: string;
+    disputeTx?: string;
+  }) {
+    this.body = data.body;
+    this.type = data.type;
+    this.timestamp = data.timestamp;
+    this.identifier = data.identifier;
+    this.requester = data.requester;
+    this.requestTx = data.requestTx;
+    this.proposer = data.proposer;
+    this.proposedValue = data.proposedValue;
+    this.proposeTx = data.proposeTx;
+    this.disputableUntil = data.disputableUntil; // should be in ms
+    this.resolvedValue = data.resolvedValue;
+    this.resolveTx = data.resolveTx;
+    this.disputeTx = data.disputeTx;
+  }
+}
+
+/**
+ * Represents a client to interact with an Optimistic Oracle and store the requests.
+ */
+export class OptimisticOracleClient {
+  protected provider: Provider;
+  protected requests: OptimisticOracleRequest[];
+  protected fetchedBlockRange: [number, number];
+
+  /**
+   * Constructs a new OptimisticOracleClient instance.
+   * @param _provider The provider used for interacting with the blockchain.
+   * @param _requests (Optional) The list of Optimistic Oracle requests.
+   */
+  constructor(_provider: Provider, _requests: OptimisticOracleRequest[] = []) {
+    this.provider = _provider;
+    this.requests = _requests;
+    this.fetchedBlockRange = [0, 0];
+  }
+
+  /**
+   * Retrieves Optimistic Oracle requests within the specified block range.
+   * @param blockRange The block range to fetch requests from.
+   * @returns A Promise that resolves to an array of OptimisticOracleRequest objects.
+   */
+  protected async fetchOracleRequests(blockRange: [number, number]): Promise<OptimisticOracleRequest[]> {
+    // TODO: Implement this.
+    blockRange;
+    return [];
+  }
+
+  /**
+   * Updates the OptimisticOracleClient instance by fetching new Oracle requests within the specified block range. Returns a new instance.
+   * @param blockRange The block range to fetch new requests from.
+   * @param existingRequests (Optional) The list of existing requests to merge with the new requests.
+   * @returns A Promise that resolves to a new OptimisticOracleClient instance with updated requests.
+   */
+  async updateWithBlockRange(
```

Comment:
nit: not sure we can make this a base class implementation since it wouldn't know which type of OptimisticOracleClient to create, right?

Similarly, if this isn't a base class implementation, it might also make sense to have fetchOracleRequests just be a private static method defined in each derived class rather than inherited, although ultimately up to you!

The way I was thinking about it (which you are free to disagree with!) was that the base class would mainly define the _read_ methods, and all the other methods could be implementation specific to the derived classes, just to keep the base interface as simple as possible for a user.

Reply:
I though about`updateWithBlockRange` as an implementation agnostic function, I think it can be in the base class given that the function `fetchOracleRequests` overridden/implemented in each class is already responsible for fetching the data for each OO flavour and convert it to the standard format. Does it make sense?